fix: client race conditions and add staleness tracking

divakarmanoj opened the pull request:
## Summary

Fix race conditions in the Client and add staleness tracking for monitoring.

## Changes

### Race Condition Fixes
- Fix data race on `closed` state using `atomic.Bool`
- Add `RWMutex` for thread-safe default client access
- Add `ticker.Stop()` to prevent resource leak

### New Features
- Add staleness tracking with `RefreshStatus` struct
- Add `IsHealthy()` - check if client is healthy
- Add `IsClosed()` - check if client is closed
- Add `GetRefreshStatus()` - get detailed refresh status
- Add `SetDefaultClient()` - explicitly set default client
- Add nil checks for default client in global functions

### Tests
- Add race condition tests for Close/GetConfig
- Add tests for staleness tracking
- Add concurrent access tests for default client
- Add benchmarks for mutex overhead (~5ns per call)

## Test plan

- [x] All tests pass with `-race` flag
- [x] Benchmarks show negligible mutex overhead

divakarmanoj committed:
fix: client race conditions and add staleness tracking

- Fix data race on closed state using atomic.Bool
- Add RWMutex for thread-safe default client access
- Add staleness tracking with RefreshStatus struct
- Add IsHealthy(), IsClosed(), GetRefreshStatus() methods
- Add SetDefaultClient() for explicit default client management
- Add nil checks for default client in global functions
- Add ticker.Stop() to prevent resource leak

Tests:
- Add race condition tests for Close/GetConfig
- Add tests for staleness tracking
- Add concurrent access tests for default client
- Add benchmarks for mutex overhead measurement

Cidyuk commented on client/client.go:
```diff
@@ -71,63 +108,172 @@ func refresh(ctx context.Context, client *Client) {
+		return false
+	}
+	status := c.GetRefreshStatus()
+	return !status.IsStale && status.LastRefreshErr == nil
```
should we use only isStale here? as `LastRefreshErr` may appear fo a brief moment of time (like s3 went down for 5 seconds) and instead of serving our config for 5 seconds we will just restart the pod over and over again and return potential 503 to the getConfig calls

divakarmanoj replied:
makes sense fixed

divakarmanoj committed:
fix: IsHealthy() only checks staleness, not last refresh error

Transient errors (e.g., brief S3 outage) shouldn't mark client as
unhealthy if data is still fresh. Use GetRefreshStatus().LastRefreshErr
for alerting/monitoring purposes instead.
